lib/middleware/karma.js

```diff
@@ -164,10 +162,18 @@ function createKarmaMiddleware (
           const scriptTags = []
           for (const file of files.included) {
             let filePath = file.path
-            const fileType = file.type || path.extname(filePath).substring(1)
+            const fileType = file.type || file.detectType()
 
-            if (helper.isDefined(fileType) && !FILE_TYPES.includes(fileType)) {
-              log.warn(`Invalid file type (${fileType}), defaulting to js.`)
+            if (!FILE_TYPES.includes(fileType)) {
+              if (file.type == null) {
+                log.warn(
+                  `Unable to determine file type from the file extension, defaulting to js.\n` +
+                  `  To silence the warning specify a valid type for ${file.originalPath} in the configuration file.\n` +
+                  `  See http://karma-runner.github.io/latest/config/files.html`
+                )
+              } else {
+                log.warn(`Invalid file type (${file.type}), defaulting to js.`)
```

Comment:
I just think the message may be confusing in the case `file.type === ''`:
  Invalid file type (), defaulting to js

Reply:
It should be `Invalid file type (empty string), defaulting to js.` now.